consensus/safety-rules/src/local.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{ConsensusState, Error, SafetyRules, TSafetyRules};
+use consensus_types::{
+    block::Block, block_data::BlockData, common::Payload, quorum_cert::QuorumCert,
+    timeout::Timeout, vote::Vote, vote_proposal::VoteProposal,
+};
+use libra_crypto::{
+    ed25519::{Ed25519PublicKey, Ed25519Signature},
+    hash::CryptoHash,
+    traits::Signature,
+};
+use libra_types::epoch_change::EpochChangeProof;
+use std::sync::{Arc, RwLock};
+
+pub struct LocalService<T> {
+    safety_rules: SafetyRules<T>,
+    pubkey: Ed25519PublicKey,
+}
+
+impl<T: Payload> LocalService<T> {
+    pub fn new(safety_rules: SafetyRules<T>, pubkey: Ed25519PublicKey) -> Self {
+        Self {
+            safety_rules,
+            pubkey,
+        }
+    }
+}
+
+/// A local interface into SafetyRules. Constructed in such a way that the container / caller
+/// cannot distinguish this API from an actual client/server process without being exposed to
+/// the actual container instead the caller can access a Box<dyn TSafetyRules>.
+pub struct LocalClient<T> {
+    internal: Arc<RwLock<LocalService<T>>>,
+}
+
+impl<T: Payload> LocalClient<T> {
+    pub fn new(internal: Arc<RwLock<LocalService<T>>>) -> Self {
+        Self { internal }
+    }
+}
+
+impl<T: Payload> TSafetyRules<T> for LocalClient<T> {
+    fn consensus_state(&mut self) -> Result<ConsensusState, Error> {
+        self.internal
+            .write()
+            .unwrap()
+            .safety_rules
+            .consensus_state()
+    }
+
+    fn initialize(&mut self, proof: &EpochChangeProof) -> Result<(), Error> {
+        self.internal
+            .write()
+            .unwrap()
+            .safety_rules
+            .initialize(proof)
+    }
+
+    fn update(&mut self, qc: &QuorumCert) -> Result<(), Error> {
+        self.internal.write().unwrap().safety_rules.update(qc)
+    }
+
+    fn construct_and_sign_vote(
+        &mut self,
+        vote_proposal: &VoteProposal<T>,
```

Comment:
signature should be within vote proposal
it should be optional
we should configure whether or not we expect signatures via the safety rules config and executor config respectively